Automatic simplification of x * 0.0 and x ** 0.0

isuruf opened the pull request:
Now `x * 0.0 = 1.0` and `x**0.0 = 1.0`

certik commented:
What exactly is the difference between is_zero and is_exact_zero?

I think it looks good, we need to also test our benchmarks.

isuruf commented:
`is_exact_zero` means it is exactly zero. For ex: `integer(0)`. `is_zero` is more relaxed. For eg: `real_double(0)`

certik commented:
Can you merge this with master? Let's make sure things work on AppVeyor. Otherwise I think I am fine with it.

isuruf commented:
I rebase against master. Let me run the benchmarks first and see if there's any slowdown

isuruf commented:
It seems there's no effect on the benchmarks. (Tried `expand2` with 10 runs and lowest was 1152ms vs the master 1153ms)

certik commented on symengine/complex.h:
```diff
@@ -64,15 +64,14 @@ class Complex : public Number {
+    virtual bool is_exact_zero() const { return false; }
     //! \return `true` if `0`
-    virtual bool is_zero() const { return ((this->real_ == 0) && (this->imaginary_ == 0)); }
+    virtual bool is_zero() const { return false; }
```
So a complex number can never be 0? How does that work?

isuruf replied:
If imaginary part is zero we return a rational or an integer. These checks are only valid for Complex numbers that are not in canonical form

certik replied:
I see. Can you put the check into an assert statement, that it is not zero? And add a comment that it cannot be zero, since it is in canonical form? That way it will be self documenting and also a bit safer in case something happened with the canonicalization.

isuruf replied:
Debug checks are already there at https://github.com/sympy/symengine/blob/8dd550ba20c0a13fb5a3e34c197c10bfaf79ae8d/symengine/complex.cpp#L20

I'll add a comment

certik replied:
I know there are there, but if we later decide to change it, then these routines like `is_zero` start failing. I think it's a good habit to check these things at the point when they are needed as well. Though you are right, that we mostly do not check what we assert in the canonicalization again. So either is fine with me. Thanks for adding the comment.

certik commented on symengine/complex.h:
```diff
@@ -64,15 +64,14 @@ class Complex : public Number {
+    virtual bool is_zero() const { return false; }
     //! \return `true` if `1`
-    virtual bool is_one() const { return ((this->real_ == 1) && (this->imaginary_ == 0)); }
+    virtual bool is_one() const { return false; }
```
And never 1?

isuruf replied:
If it is 1, then it should be an integer, not a Complex

certik commented on symengine/number.h:
```diff
@@ -15,7 +15,9 @@ class Evaluate;
+    //! \return true if exactly equal to `0`
+    virtual bool is_exact_zero() const = 0;
+    //! \return true if equal to `0`
     virtual bool is_zero() const = 0;
```
Can you describe in the documentation here what the difference between `is_exact_zero()` and `is_zero()` is and when one should use each?

Also, we need to add tests for these two functions. We only test this indirectly so far.

isuruf replied:
Sure. I'll add documentation and tests.

certik commented on symengine/rational.h:
```diff
@@ -45,21 +45,17 @@ class Rational : public Number {
+    virtual bool is_zero() const { return false; }
     //! \return `true` if `1`
-    virtual bool is_one() const { return this->i == 1; }
+    virtual bool is_one() const { return false; }
```
Why is `is_one()` and `is_zero()` false here always?

isuruf replied:
Same as before. It should be an integer in that case.

isuruf commented:
@certik, I realized that instead of defining `is_exact_zero` function in each number class, I can do `is_zero() && is_exact()`. Is this okay given the fact that it takes two virtual function calls?

certik commented:
I am still confused what these two functions really mean. For example, given the amount of change you had to do, wouldn't it be better to rename `is_exact_zero()->is_zero` and `is_zero() -> is_inexact_zero()` or something. What does `is_exact()` do?

isuruf commented:
`is_exact()` was a function that I introduced when adding automatic simplification. 

For example `sin(1)` is kept as it is, while `sin(1.0)` is automatically simplified. The decision to automatically simplify is based on whether the arg of the function is an inexact number or not.

There are too many things I've changed in this PR. I'll send separate ones.

certik commented:
I see. To answer your question, it think it's fine to call the two functions, but we should run benchmarks.

isuruf commented:
Closing in favour of #565
